src/PostgREST/App.hs

```diff
@@ -9,430 +9,664 @@ Some of its functionality includes:
+  do
+    body <- lift $ Wai.strictRequestBody req
+
+    dbStructure <-
+      case maybeDbStructure of
+        Just dbStructure ->
+          return dbStructure
+
+        Nothing ->
+          throwError Error.ConnectionLostError
+
+    apiRequest <-
+      liftEither . mapLeft Error.ApiRequestError $
+        Req.userApiRequest
+          (Config.configDbSchemas conf)
+          (Config.configDbRootSpec conf)
+          dbStructure
+          req
+          body
+
+    -- The JWT must be checked before touching the db
+    jwtClaims <- Auth.jwtClaims conf (toS $ Req.iJWT apiRequest) time
+
+    let
+      reqContentTypes = requestContentTypes conf apiRequest
+      acceptContentType = Req.iAccepts apiRequest
+
+    contentType <-
+      case Req.mutuallyAgreeable reqContentTypes acceptContentType of
+        Just ct ->
+          return ct
+
+        Nothing ->
+          throwError . Error.ContentTypeError $ map Types.toMime acceptContentType
+
+    let
+      context apiReq =
+        RequestContext conf dbStructure apiReq contentType
+
+      handleReq apiReq =
+        handleRequest (context apiReq)
+
+    runDbHandler pool (txMode apiRequest) jwtClaims $
+      Middleware.optionalRollback conf apiRequest $
+        Middleware.runPgLocals conf jwtClaims handleReq apiRequest
+
+runDbHandler :: SQL.Pool -> SQL.Mode -> Auth.JWTClaims -> DbHandler a -> Handler IO a
+runDbHandler pool mode jwtClaims handler =
+  do
+    dbResp <-
+      lift . SQL.use pool . SQL.transaction SQL.ReadCommitted mode $
+        runExceptT handler
+
+    resp <-
+      liftEither . mapLeft Error.PgErr $
+        mapLeft (Error.PgError $ Auth.containsRole jwtClaims) dbResp
+
+    liftEither resp
+
+handleRequest :: RequestContext -> DbHandler Wai.Response
+handleRequest context@(RequestContext _ dbStructure apiRequest _) =
+  case (Req.iAction apiRequest, Req.iTarget apiRequest) of
+    (Req.ActionRead headersOnly, Req.TargetIdent identifier) ->
+      handleRead headersOnly identifier context
+    (Req.ActionCreate, Req.TargetIdent identifier) ->
+      handleCreate identifier context
+    (Req.ActionUpdate, Req.TargetIdent identifier) ->
+      handleUpdate identifier context
+    (Req.ActionSingleUpsert, Req.TargetIdent identifier) ->
+      handleSingleUpsert identifier context
+    (Req.ActionDelete, Req.TargetIdent identifier) ->
+      handleDelete identifier context
+    (Req.ActionInfo, Req.TargetIdent identifier) ->
+      handleInfo identifier dbStructure
+    (Req.ActionInvoke invMethod, Req.TargetProc proc _) ->
+      handleInvoke invMethod proc context
+    (Req.ActionInspect headersOnly, Req.TargetDefaultSpec tSchema) ->
+      handleOpenApi headersOnly tSchema context
+    _ ->
+      throwError Error.NotFound
+
+handleRead :: Bool -> QualifiedIdentifier -> RequestContext -> DbHandler Wai.Response
+handleRead headersOnly identifier context@(RequestContext conf dbStructure apiRequest contentType) =
```

Comment:
```suggestion
handleRead headersOnly identifier context@(RequestContext AppConfig{..} dbStructure ApiRequest{..} contentType) =
```

I think for those two, this would be a pattern that makes sense across the whole file. I commented a lot of places already - but it just keeps coming up again, and the more I comment it the more I like it. It just removes boilerplate.